src/rust/lib/graph-editor/src/component/node/port/output.rs

```diff
@@ -0,0 +1,682 @@
+            }
+        }
+    }
+}
+
+
+
+// =============================
+// === Port Frp Setup Helper ===
+// =============================
+
+/// Helper struct to pass the required FRP endpoints to set up the FRP of a port shape view.
+#[derive(Clone,CloneRef,Debug)]
+struct PortFrp {
+    network                      : frp::Network,
+
+    port_mouse_over              : frp::Source<PortId>,
+    port_mouse_out               : frp::Source<PortId>,
+    port_mouse_down              : frp::Source<PortId>,
+
+    hide_all                     : frp::Stream<()>,
+    activate_ports_with_selected : frp::Stream<PortId>
+}
+
+/// Set up the FRP system for a ShapeView of a shape that implements the PortShapeApi.
+///
+/// This allows us to use the same setup code for bot the `multi_port_area::Shape` and the
+/// `single_port_area::Shape`.
+fn init_port_frp<Shape:PortShapeApi+CloneRef+'static>
+(view:&component::ShapeView<Shape>, port_id:PortId,frp:PortFrp) {
+    let PortFrp { network,port_mouse_over,port_mouse_out,port_mouse_down,
+        hide_all,activate_ports_with_selected } = frp;
+
+    let shape        = &view.shape;
+    let port_size    = Animation::<f32>::new(&network);
+    let port_opacity = Animation::<f32>::new(&network);
+
+    frp::extend! { network
+
+            // === Mouse Event Handling == ///
+
+            eval_ view.events.mouse_over(port_mouse_over.emit(port_id));
+            eval_ view.events.mouse_out(port_mouse_out.emit(port_id));
+            eval_ view.events.mouse_down(port_mouse_down.emit(port_id));
+
+
+             // === Animation Handling == ///
+
+             eval port_size.value    ((size) shape.set_grow(*size));
+             eval port_opacity.value ((size) shape.set_opacity(*size));
+
+
+            // === Visibility and Highlight Handling == ///
+
+             def _hide_all = hide_all.map(f_!([port_size,port_opacity]{
+                 port_size.set_target_value(0.0);
+                 port_opacity.set_target_value(0.0);
+             }));
+
+            // Through the provided ID we can infer whether this port should be highlighted.
+            is_selected      <- activate_ports_with_selected.map(move |id| *id == port_id);
+            show_normal      <- activate_ports_with_selected.gate_not(&is_selected);
+            show_highlighted <- activate_ports_with_selected.gate(&is_selected);
+
+            eval_ show_highlighted ([port_opacity,port_size]{
+                port_opacity.set_target_value(1.0);
+                port_size.set_target_value(HIGHLIGHT_SIZE);
+            });
+
+            eval_ show_normal ([port_opacity,port_size]
+                port_opacity.set_target_value(0.5);
+                port_size.set_target_value(BASE_SIZE);
+            );
+        }
+}
+
+
+
+// ===========
+// === Frp ===
+// ===========
+
+/// Id of a specific port inside of `OutPutPortsData`.
+#[derive(Clone,Copy,Default,Debug,Eq,PartialEq)]
+pub struct PortId {
```

Comment:
I think a better name would be `PortIndex` instead. Correct me if Im wrong, but these are indexes of ports starting from left right?

Reply:
Internally yes, but that should not matter for the API user. They should just assume that this struct identifies a Port in an OutputPort struct. So I'd rather keep the  `Id` instead of `Index` suffix.